v-next/hardhat-zod-utils/test/index.ts

```diff
@@ -32,3 +41,146 @@ describe("unionType", () => {
     assertParseResult(union.safeParse(undefined), "Expected error message");
   });
 });
+
+describe("conditionalUnionType", () => {
+  describe("Conditions evaluation", () => {
+    it("should return the first type that matches", async () => {
+      const shouldUseString = conditionalUnionType(
+        [
+          [(data) => typeof data === "string", z.string()],
+          [(data) => typeof data === "string", z.string().url()],
+        ],
+        "No match",
+      );
+
+      // Both conditions match, but we only use the first one
+      assert.equal(shouldUseString.safeParse("asd").success, true);
+
+      const shouldUseUrl = conditionalUnionType(
+        [
+          [(data) => typeof data === "string", z.string().url()],
+          [(data) => typeof data === "string", z.string()],
+        ],
+        "No match",
+      );
+
+      assertParseResult(shouldUseUrl.safeParse("asd"), "Invalid url");
+    });
+
+    it("should return the provided error message if no condition matches", async () => {
+      const shouldUseString = conditionalUnionType(
+        [
+          [(data) => typeof data === "string", z.string()],
+          [(data) => typeof data === "string", z.string().url()],
+        ],
+        "No match",
+      );
+
+      // Both conditions match, but we only use the first one
+      assertParseResult(shouldUseString.safeParse(123), "No match");
+    });
+  });
+
+  describe("Zod issues paths", () => {
+    it("should have an empty path when nothing matches in as a top-level type", async () => {
+      const shouldUseString = conditionalUnionType(
+        [
+          [(data) => typeof data === "string", z.string()],
+          [(data) => typeof data === "string", z.string().url()],
+        ],
+        "No match",
+      );
+
+      assertParseResult(shouldUseString.safeParse(123), "No match", []);
+    });
+
+    it("Should have the path to the nested error if a condition matches", () => {
+      const shouldUseString = conditionalUnionType(
+        [
+          [(data) => typeof data === "object", z.object({ foo: z.string() })],
```

Comment:
not that it matters too much here, but I'd still use isObject in tests, as it could be viewed or copied by another dev later on.

Reply:
yep